content/posts/faster-rfft-python/index.md

```diff
@@ -0,0 +1,386 @@
+
+
+### Timing
+Let's start by preparing a script that can get the timing of RFFT computation. This timing function will be used  to evaluate the performance of different RFFT implementations.
+
+```python
+def _get_timing_single(n, rfft_caller, timeout=5.0, iter_max=1000000):
+    """
+    For a single input size n, return the time taken to compute RFFT using rfft_caller
+    """
+    pyfftw.forget_wisdom()  # clean up any previous plans
+    T = np.random.rand(n)
+
+    R = rfft_caller(T) # dummy to create wisdom
+    np.testing.assert_allclose(R, np.fft.rfft(T))  # verify correctness
+
+    total_time = 0.0
+    count = 0
+    while total_time < timeout and count < iter_max:
+        start_time = time.perf_counter()
+        rfft_caller(T)
+        total_time += time.perf_counter() - start_time
+        count += 1
+
+    return total_time / count
+
+
+def get_timing(n_values, rfft_caller, timeout=5.0, iter_max=1000000, verbose=True):
+    """
+    For multiple input sizes in n_values, return the time taken to compute RFFT using rfft_caller
+    """
+    timing = np.full(len(n_values), -1.0, dtype='float64')
+    for i, n in enumerate(n_values):
+        timing[i] = _get_timing_single(n, rfft_caller, timeout=timeout, iter_max=iter_max)
+        if verbose:
+            print(f"log2(n) --> {int(np.log2(n))}, time: {timing[i]:.6f} sec", flush=True)
+    return timing
+```
+
+`_get_timing_single` measures the average time to perform an RFFT on an array of size `n` using the supplied `rfft_caller`. The function `get_timing` simply extends this logic to multiple input sizes listed in n_values.
+
+In `_get_timing_single`, we first clear any previously stored FFTW plans (known as “wisdom”) using `pyfftw.forget_wisdom()`. If you’re not familiar with FFTW wisdom, here’s a quick summary: For a new transformation for a new input size, FFTW may spend time exploring different algorithms to find the fastest strategy in the first call. The result of this search, "plan", is essentially a performance recipe that allows future FFTs of the same input size and type to skip the search entirely. This wisdom is cached by `pyfftw` internally. Running `pyfftw.forget_wisdom()` ensures that we start fresh without any cached plans. For a deeper explanation, see [this short blog post](https://the-data-philomath.netlify.app/posts/pyfftw-wisdom/).
+
+Next, we generate a random input array `T` of length `n`. We then perform a dummy RFFT call using `rfft_caller(T)`, mainly to allow FFTW to build the optimal plan (i.e., compute wisdom) ahead of timing. We also validate correctness by comparing the output to NumPy’s `np.fft.rfft` via `np.testing.assert_allclose`. This step is important because it ensures that we are timing a correct implementation.
```

Comment:
> We then perform a dummy RFFT call using `rfft_caller(T)`, mainly to allow FFTW to build the optimal plan (i.e., compute wisdom) ahead of timing

It is important to help users understand the hidden assumption here! The assumption is that  we know, in advance, the size/type of inputs. Therefore, we can pre-compute plan, and save it for later to avoid computing that again. That's why its timing is excluded. However, if the application needs to apply fft on new, different (unforeseen) input sizes, the planning time can become an important factor.